### ISSUE TITLE
Xterm.js overwrites the current line on overflow with fit addon

### ISSUE THREAD

**Issue by ashblue** [xtermjs/xterm.js#4434]
Xterm.js handles line overflow by writing overflowing characters onto another line. But when Xterm.js runs the fit addon it overwrites the same line.

This bug is purely cosmetic. But it can be frustrating when handling long database or CURL commands. It would be nice if the fit addon provided an option to declare how to handle overflow.

Expected behavior:
<img width="853" alt="image" src="https://user-images.githubusercontent.com/643307/224130355-f2c4ec57-0383-4465-8d3a-06d9175f4b2e.png">

What we're seeing:
<img width="1096" alt="image" src="https://user-images.githubusercontent.com/643307/224131129-31def18e-a600-462f-8279-4a8a07805b7c.png">

## Details
- Browser and browser version:  Chrome 110.0.5481.177
- OS version: Mac 12.6.1
- xterm.js version: `{ "xterm": "~5.1.0", "xterm-addon-fit": "~0.7.0", }`

### Steps to reproduce

1. Call the fit addon
2. Type enough characters that it causes overflow
3. Notice the characters start to overwrite each other

**Comment by saowu** [comment]
I have the same problem

**Comment by jerch** [comment]
Plz see my comment here https://github.com/xtermjs/xterm.js/issues/4419#issuecomment-1463687838

**Comment by 18578242152** [comment]
> I met this in bash and solved by triggering an terminal.resize when websocket connect #4419
> 
> ```
>     const listenResize = function () {
>       terminal.loadAddon(fitAddon);
>       fitAddon.fit();
>       const msg = { type: 'resize', rows: terminal.rows, cols: terminal.cols };
>       socket.send(JSON.stringify(msg));
>     };
> 
>     socket.onopen = function () {
>       console.log('socket open');
>       listenResize();
>     };
> ```

Hello, why doesn't it work
<img width="547" alt="Screen Shot 2023-03-23 at 5 43 26 PM" src="https://user-images.githubusercontent.com/107911122/227164575-4f131207-7910-4e32-b700-a68d22b9b149.png">

**Comment by saowu** [comment]
Please try to set env TERM=dummy, which can barely minimize the impact

![image](https://user-images.githubusercontent.com/33569619/227436321-5bbee15c-5bf6-4765-95c0-4e2ba856a175.png)

**Comment by jerch** [comment]
No, TERM env var has nothing to do with resize handling.

_What is TERM var for?_
TERM var is the key to an entry of the terminfo database (see [terminfo(5)](https://man7.org/linux/man-pages/man5/terminfo.5.html)). It tells applications, which type of terminal sequences are supposed to work at the controlling terminal. Some shells and all ncurses based application make use of this to do more fancy output manipulations.

By setting TERM=dumb you basically told apps using terminfo to treat the terminal as a dumb typewriter not supporting any higher sequences, e.g. it turns off colors, cursor jumps, tab support etc. It also only affects terminfo aware applications. Thats most likely not what you want:
```
infocmp dumb 
#       Reconstructed via infocmp from file: /lib/terminfo/d/dumb
dumb|80-column dumb tty,
        am,
        cols#80,
        bel=^G, cr=\r, cud1=\n, ind=\n,
```
As you can see, dumb is only supposed to understand NL, CR and BEL, with an assumed width of 80 cols. No height notion. That pretty much only resembles a real mechanical typewriter without TAB support. I know it is hard to follow if you never touched such a machine yourself, but the TTY + terminal interface are derived from this basic typewriter idea.

_Whats the matter with terminal resizes?_

In the very early days glassy terminals were not resizable at all, they had a fixed character grid size, but would differ between vendors. Thats where initial size notions in terminfo comes from. But soon vendors emulated other modes as well and supported different grid sizes (e.g. VT100 already knew 2 grid resolutions). With the upcoming of GUI driven terminal emulators (TE) the grid size was not bound anymore to device restrictions, in fact the TE window was freely resizable. Thats a problem for the underlying system, that started very rigid in its size assumptions with partially hardcoded metrics. So ppl extended the TTY/PTY system with a size notion, that can be altered on the fly, together with SIGWINCH to signal the size change to the running process.

And thats what we have today - the PTY interface is mainly about 2 aspects:
- IO data sink (with different [termios(3)](https://man7.org/linux/man-pages/man3/termios.3.html) operation modes), meant to be changed from app side
- size notion (see [tty_ioctl(4)](https://linux.die.net/man/4/tty_ioctl)), meant to be altered from TE side

The workflow behind the latter is the following:
- TE changes grid size from 80x24 --> 100x50 by a user action.
  With xterm.js that can be achieved by resizing the terminal container div and calling `FitAddon.fit()`. The addon will recalculate the new grid size based on the new div size. _Thats all the fit addon does._
- The changed grid size gets announced to the PTY.
  For xterm.js this means, you have to grab the new size values and transport them somehow to your backend. In the backend use the appropriate method to update PTY size. Most PTY bridges expose a suitable method for this, eg. with `node-pty` call `your_pty_object.resize(cols, rows)`.
- Kernel updates internal size struct and triggers SIGWINCH to running processes.
  Finally SIGWINCH will be issued, so applications have a chance to get notified about the size chance. Most advanced console editors and ncurses apps will redraw itself on SIGWINCH. Also shells read the PTY size to do their fancy line manipulations. Thats the point, where a size mismatch will create the OP issue.

I admit that this handling is pretty awful, but it is what you get, when ~50ys of incremental evolution without cutting old ropes shape things.

**Comment by saowu** [comment]
是的，但是我调整env TERM后，覆盖问题不存在了，其他并无明显影响。




吴延博
***@***.***




------------------ Original ------------------
From: jerch ***@***.***&gt;
Date: Fri,Mar 24,2023 9:06 PM
To: xtermjs/xterm.js ***@***.***&gt;
Cc: 吴延博 ***@***.***&gt;, Comment ***@***.***&gt;
Subject: Re: [xtermjs/xterm.js] Xterm.js overwrites the current line onoverflow with fit addon (Issue #4434)





 
No, TERM env var has nothing to do with resize handling.
 
What is TERM var for?
 TERM var is the key to an entry of the terminfo database (see terminfo(5)). It tells applications, which type of terminal sequences are supposed to work at the controlling terminal. Some shells and all ncurses based application make use of this to do more fancy output manipulations.
 
By setting TERM=dumb you basically told apps using terminfo to treat the terminal as a dumb typewriter not supporting any higher sequences, e.g. it turns off colors, cursor jumps, tab support etc. It also only affects terminfo aware applications. Thats most likely not what you want:
 infocmp dumb  #       Reconstructed via infocmp from file: /lib/terminfo/d/dumb dumb|80-column dumb tty,         am,         cols#80,         bel=^G, cr=\r, cud1=\n, ind=\n,  
As you can see, dumb is only supposed to understand NL, CR and BEL, with an assumed width of 80 cols. No height notion. That pretty much only resembles a real mechanical typewriter without TAB support. I know it is hard to follow if you never touched such a machine yourself, but the TTY + terminal interface are derived from this basic typewriter idea.
 
Whats the matter with terminal resizes?
 
In the very early days glassy terminals were not resizable at all, they had a fixed character grid size, but would differ between vendors. Thats where initial size notions in terminfo comes from. But soon vendors emulated other modes as well and supported different grid sizes (e.g. VT100 already knew 2 grid resolutions). With the upcoming of GUI driven terminal emulators (TE) the grid size was not bound anymore to device restrictions, in fact the TE window was freely resizable. Thats a problem for the underlying system, that started very rigid in its size assumptions with partially hardcoded metrics. So ppl extended the TTY/PTY system with a size notion, that can be altered on the fly, together with SIGWINCH to signal the size change to the running process.
 
And thats what we have today - the PTY interface is mainly about 2 aspects:
  
IO data sink (with different termios(3) operation modes), meant to be changed from app side
 
size notion (see tty_ioctl(4)), meant to be altered from TE side
  
The workflow behind the latter is the following:
  
TE changes grid size from 80x24 --&gt; 100x50 by a user action.
 With xterm.js that can be achieved by resizing the terminal container div and calling FitAddon.fit(). The addon will recalculate the new grid size based on the new div size. Thats all the fit addon does.
 
The changed grid size gets announced to the PTY.
 For xterm.js this means, you have to grab the new size values and transport them somehow to your backend. In the backend use the appropriate method to update PTY size. Most PTY bridges expose a suitable method for this, eg. with node-pty call your_pty_object.resize(cols, rows).
 
Kernel updates internal size struct and triggers SIGWINCH to running processes.
 Finally SIGWINCH will be issued, so applications have a chance to get notified about the size chance. Most advanced console editors and ncurses apps will redraw itself on SIGWINCH. Also shells read the PTY size to do their fancy line manipulations. Thats the point, where a size mismatch will create the OP issue.
  
I admit that this handling is pretty awful, but it is what you get, when ~50ys of incremental evolution without cutting old ropes shape things.
